### PR TITLE
WIP: Support atomics

### PR DESCRIPTION
**Pull request by dylanmckay**
This adds atomics support to AVR-LLVM. Based of @shepmaster's original work.

Fix for #197

### CODE REVIEW

**Comment by dylanmckay** [comment]
I also had trouble with gluing all the nodes together. Have asked for help on the [llvm-dev list](http://lists.llvm.org/pipermail/llvm-dev/2016-June/100663.html).

File: lib/Target/AVR/AVRISelLowering.cpp [hunk 1]
```diff
@@ -157,6 +157,11 @@ AVRTargetLowering::AVRTargetLowering(AVRTargetMachine &tm)
     // improvements in how we treat 16-bit "registers" to be feasible.
   }
 
+  for (MVT VT : MVT::integer_valuetypes()) {
```
Comment (shepmaster) [inline lib/Target/AVR/AVRISelLowering.cpp]: I was being lazy originally. I don't *think* that an `i8` needs to have the atomic wrapping as the CPU cannot break it up into pieces.

File: lib/Target/AVR/AVRISelLowering.cpp [hunk 2]
```diff
@@ -366,6 +372,42 @@ SDValue AVRTargetLowering::LowerDivRem(SDValue Op, SelectionDAG &DAG) const {
...
+// Save status register which contains the interrupt
+// state. Unconditionally disable interrupts, do our work, then
+// potentially re-enable interrupts.
+SDValue AVRTargetLowering::LowerAtomic(SDValue AtomicOp, SelectionDAG &DAG) const {
+  SDLoc DL(AtomicOp);
+
+  // Store `SREG`
+  SDValue StatusReg = DAG.getCopyFromReg(DAG.getEntryNode(), SDLoc(AtomicOp), AVR::SREG, MVT::i8);
+
+  // Disable interrupts (`clr` is equivalent to `bclr 7`).
```
Comment (shepmaster) [inline lib/Target/AVR/AVRISelLowering.cpp]: Pedantic: `clr` should be `cli`.

**Comment by dylanmckay** [comment]
LLVMDev suggests that the only _real_ way to do this is in the machine instruction stage.

Even if we managed to glue all the nodes together, the pipeline would likely insert spills and stuff inbetween.

**Pull request closed without merging**